plugins/redfish/fu-plugin-redfish.c

```diff
@@ -8,69 +8,179 @@
+	}
+	fu_redfish_backend_set_username (data->backend, split[0]);
+	fu_redfish_backend_set_password (data->backend, split[1]);
+	return TRUE;
+}
+
+static gboolean
+fu_redfish_plugin_discover_smbios_table (FuPlugin *plugin, GError **error)
+{
+	FuPluginData *data = fu_plugin_get_data (plugin);
+	FuContext *ctx = fu_plugin_get_context (plugin);
+	const gchar *smbios_data_fn;
+	g_autofree gchar *hostname = NULL;
+	g_autoptr(FuRedfishSmbios) redfish_smbios = fu_redfish_smbios_new ();
+	g_autoptr(GBytes) smbios_data = NULL;
+
+	/* is optional if not in self tests */
+	smbios_data_fn = g_getenv ("FWUPD_REDFISH_SMBIOS_DATA");
+	if (smbios_data_fn != NULL) {
+		smbios_data = fu_common_get_contents_bytes (smbios_data_fn, error);
+		if (smbios_data == NULL)
+			return FALSE;
+	} else {
+		smbios_data = fu_context_get_smbios_data (ctx, REDFISH_SMBIOS_TABLE_TYPE);
+		if (smbios_data == NULL)
+			return TRUE;
+	}
+	if (!fu_firmware_parse (FU_FIRMWARE (redfish_smbios),
+				smbios_data, FWUPD_INSTALL_FLAG_NONE, error)) {
+		g_prefix_error (error, "failed to parse SMBIOS table entry type 42: ");
+		return FALSE;
+	}
+
+	/* get IP, falling back to hostname, then MAC, then VID:PID */
+	hostname = g_strdup (fu_redfish_smbios_get_ip_addr (redfish_smbios));
+	if (hostname == NULL)
+		hostname = g_strdup (fu_redfish_smbios_get_hostname (redfish_smbios));
+	if (hostname == NULL) {
+		const gchar *mac_addr = fu_redfish_smbios_get_mac_addr (redfish_smbios);
+		if (mac_addr != NULL) {
+			hostname = fu_redfish_network_ip_for_mac_addr (mac_addr, error);
+			if (hostname == NULL)
+				return FALSE;
+		}
+	}
+	if (hostname == NULL) {
+		guint16 vid = fu_redfish_smbios_get_vid (redfish_smbios);
+		guint16 pid = fu_redfish_smbios_get_pid (redfish_smbios);
+		if (vid != 0x0 && pid != 0x0) {
+			hostname = fu_redfish_network_ip_for_vid_pid (vid, pid, error);
+			if (hostname == NULL)
+				return FALSE;
+		}
+	}
+	if (hostname == NULL) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_INVALID_FILE,
+				     "no hostname");
+		return FALSE;
+	}
+	fu_redfish_backend_set_hostname (data->backend, hostname);
+	fu_redfish_backend_set_port (data->backend, fu_redfish_smbios_get_port (redfish_smbios));
+	return TRUE;
+}
+
 gboolean
 fu_plugin_startup (FuPlugin *plugin, GError **error)
 {
 	FuPluginData *data = fu_plugin_get_data (plugin);
-	FuContext *ctx = fu_plugin_get_context (plugin);
-	gboolean ca_check;
+	g_autofree gchar *ca_check_str = NULL;
+	g_autofree gchar *password = NULL;
 	g_autofree gchar *redfish_uri = NULL;
-	g_autoptr(GBytes) smbios_data = NULL;
+	g_autofree gchar *username = NULL;
+	g_autoptr(GError) error_uefi = NULL;
 
 	/* optional */
-	smbios_data = fu_context_get_smbios_data (ctx, REDFISH_SMBIOS_TABLE_TYPE);
+	if (!fu_redfish_plugin_discover_smbios_table (plugin, error))
+		return FALSE;
+	if (!fu_redfish_plugin_discover_uefi_credentials (plugin, &error_uefi)) {
+		g_debug ("failed to get username and password automatically: %s",
```

Comment:
Isn't this useful to show to the user somehow?  It would be good to be able to tell whether you need to fill out a conffile or if there is a problem.

Reply:
The UEFI credentials have been deprecated apparently, and so failure to have them set isn't a real problem. It's only a problem if we can't get them from anywhere else, e.g. from the conf file or from KCS -- the latter bit I've not written yet.